### PR TITLE
feat: Added webView based myPrograms

### PR DESCRIPTION
**Pull request by omerhabib26**
### Description:

- Create Separate directories for Dashboard & MyPrograms
- Added webView-based myPrograms
- Added config dic for myPrograms
- Move URI_SCHEME to class level in config
- Update TestCases for some classes 

Ref: [Issue: 157](https://github.com/openedx/openedx-app-android/issues/157)

### CODE REVIEW

**Review by HamzaIsrar12 — CHANGES_REQUESTED** [review]
Points from manual & initial review: 

- Please implement swipe refresh for Program screen
- Program WebView: 
    1. Clicking on an enrolled course should redirect that course
    2. Clicking on an un-erolled course should redirect to Discover's course info screen
    3. Clicking on enroll button should enrol and redirect the user to that course

File: core/src/main/java/org/openedx/core/presentation/catalog/WebViewLink.kt [hunk 1]
```diff
@@ -16,7 +16,8 @@ class WebViewLink(
         ENROLL("enroll"),
         ENROLLED_PROGRAM_INFO("enrolled_program_info"),
         ENROLLED_COURSE_INFO("enrolled_course_info"),
-        COURSE("course")
+        COURSE("course"),
+        EXTERNAL("external"),// This is a link to an external site
```
Comment (HamzaIsrar12) [inline core/src/main/java/org/openedx/core/presentation/catalog/WebViewLink.kt:20]: Maybe we can remove the comment. 🤔

**Review by HamzaIsrar12 — CHANGES_REQUESTED** [review]
Features to look into:
- Swipe to refresh
- Snackbar after successful course enrollment

**Review by HamzaIsrar12 — CHANGES_REQUESTED** [review]
Minor nits only.

File: dashboard/src/main/java/org/openedx/dashboard/presentation/dashboard/DashboardViewModel.kt [hunk 2]
```diff
@@ -63,12 +66,19 @@ class DashboardViewModel(
...
         super.onCreate(owner)
         viewModelScope.launch {
-            notifier.notifier.collect {
+            courseNotifier.notifier.collect {
                 if (it is CourseDashboardUpdate) {
                     updateCourses()
                 }
             }
         }
+        viewModelScope.launch {
```
[the remark this answers is not in the document]
Comment (omerhabib26) [reply dashboard/src/main/java/org/openedx/dashboard/presentation/dashboard/DashboardViewModel.kt:75]: `viewModelScope` works for single thread operation

File: dashboard/src/main/java/org/openedx/dashboard/notifier/DashboardEvent.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package org.openedx.dashboard.notifier
+
+interface DashboardEvent
```
[the remark this answers is not in the document]
Comment (omerhabib26) [reply dashboard/src/main/java/org/openedx/dashboard/notifier/DashboardEvent.kt:3]: updated

File: app/src/main/java/org/openedx/app/MainViewModel.kt [hunk 4]
```diff
@@ -1,20 +1,53 @@
...
+
     val isDiscoveryTypeWebView get() = config.getDiscoveryConfig().isViewTypeWebView()
 
+    val isProgramTypeWebView get() = config.getProgramConfig().isViewTypeWebView()
+
+    init {
+        runBlocking {
+            _navigateToDiscovery.emit(false)
+        }
+    }
```
[the remark this answers is not in the document]
Comment (omerhabib26) [reply app/src/main/java/org/openedx/app/MainViewModel.kt:38]: removed

File: app/src/main/java/org/openedx/app/MainFragment.kt [hunk 5]
```diff
@@ -68,6 +73,15 @@ class MainFragment : Fragment(R.layout.fragment_main) {
             enableBottomBar(isBottomBarEnabled)
         }
 
+        viewLifecycleOwner.lifecycleScope.launch {
+            viewModel.navigateToDiscovery.collect { shouldNavigateToDiscovery ->
+                if (shouldNavigateToDiscovery) {
+                    binding.bottomNavView.selectedItemId = R.id.fragmentHome
+                    binding.viewPager.setCurrentItem(0, true)
```
[the remark this answers is not in the document]
Comment (omerhabib26) [reply app/src/main/java/org/openedx/app/MainFragment.kt:80]: removed

File: dashboard/src/main/java/org/openedx/dashboard/presentation/program/ProgramFragment.kt [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+                val showLoading by viewModel.showLoading.collectAsState(initial = true)
+                val enrollmentSuccess by viewModel.courseEnrollSuccess.collectAsState(initial = "")
+                var hasInternetConnection by remember {
+                    mutableStateOf(viewModel.hasInternetConnection)
+                }
+
+                LaunchedEffect(showAlert) {
+                    if (showAlert) {
+                        InfoDialogFragment.newInstance(
+                            title = context.getString(courseR.string.course_enrollment_error),
```
Comment (k1rill) [inline dashboard/src/main/java/org/openedx/dashboard/presentation/program/ProgramFragment.kt:87]: please use stringResource fun from compose here and for others strings in this fun

File: dashboard/src/main/java/org/openedx/dashboard/presentation/program/ProgramFragment.kt [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+    ) = ComposeView(requireContext()).apply {
+        setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+        setContent {
+            OpenEdXTheme {
+                val windowSize = rememberWindowSize()
+                val uiMessage by viewModel.uiMessage.collectAsState(initial = null)
+                val showAlert by viewModel.showAlert.collectAsState(initial = false)
+                val showLoading by viewModel.showLoading.collectAsState(initial = true)
+                val enrollmentSuccess by viewModel.courseEnrollSuccess.collectAsState(initial = "")
+                var hasInternetConnection by remember {
```
Comment (k1rill) [inline dashboard/src/main/java/org/openedx/dashboard/presentation/program/ProgramFragment.kt:80]: nit: not very good to use var here

**Review by k1rill — COMMENTED** [review]
only some nits

**Review by HamzaIsrar12 — DISMISSED** [review]
Approved from my side, please proceed with the merge after addressing Kirill's feedback. 🚀

File: dashboard/src/main/java/org/openedx/dashboard/presentation/program/ProgramFragment.kt (hunk 6 above)
Comment (k1rill) [reply dashboard/src/main/java/org/openedx/dashboard/presentation/program/ProgramFragment.kt:87]: here is composabale context, it is inside setContent of the compose view

File: app/src/main/java/org/openedx/app/MainViewModel.kt [hunk 8]
```diff
@@ -1,20 +1,58 @@
...
+                    try {
+                        interactor.enrollInACourse(it.courseId)
+                        notifier.send(DashboardEvent.CourseEnrolledSuccess(it.courseId))
+                    } catch (
+                        e: Exception
+                    ) {
+                        e.printStackTrace()
+                        notifier.send(DashboardEvent.CourseEnrolledError(e))
+                    }
+                }
```
Comment (volodymyr-chekyrta) [inline app/src/main/java/org/openedx/app/MainViewModel.kt:51]: It's a nice workaround to not connect the course module, but controlling it via the event bus is non-obvious and a bit confusing. We can do it in a cleaner way.
I would recommend creating an interface like `EnrollInACourseUseCase` or `EnrollInACourseInteractor` in the `core` module and implementing that interface in the `CourseInteractor`.
After that, all you need to do is simply provide/inject the `EnrollInACourseInteractor` interface into `ProgramViewModel`.

File: core/src/main/java/org/openedx/core/ui/ComposeCommon.kt [hunk 9]
```diff
@@ -383,14 +363,21 @@ fun HandleUIMessage(
...
     }
 }
 
+fun toastMessage(
+    context: Context,
+    message: String
+) {
+    Toast.makeText(context, message, Toast.LENGTH_SHORT).show()
+}
+
```
[the remark this answers is not in the document]
Comment (omerhabib26) [reply core/src/main/java/org/openedx/core/ui/ComposeCommon.kt:380]: moved to View.kt

File: app/src/main/java/org/openedx/app/MainViewModel.kt (hunk 8 above)
Comment (omerhabib26) [reply app/src/main/java/org/openedx/app/MainViewModel.kt]: updated

File: app/src/main/java/org/openedx/app/di/AppModule.kt [hunk 10]
```diff
@@ -153,4 +157,6 @@ val appModule = module {
     factory { FacebookAuthHelper() }
     factory { GoogleAuthHelper(get()) }
     factory { MicrosoftAuthHelper() }
+
+    single<EnrollInCourseInteractor> { get<CourseInteractor>() }
```
Comment (volodymyr-chekyrta) [inline app/src/main/java/org/openedx/app/di/AppModule.kt:161]: Do we need to keep this dependency as a singleton?
CourseInteractor, by nature, does not keep any state or data, so it seems there is no reason to keep EnrollInCourseInteractor as singleton.

**Review by k1rill — DISMISSED** [review]

File: app/src/main/java/org/openedx/app/di/AppModule.kt (hunk 10 above)
Comment (omerhabib26) [reply app/src/main/java/org/openedx/app/di/AppModule.kt]: updated

**Review by volodymyr-chekyrta — APPROVED** [review]

**Review by k1rill — DISMISSED** [review]

**Review by HamzaIsrar12 — APPROVED** [review]

**Review by farhan-arshad-dev — APPROVED** [review]

**Pull request merged**